File: packages/domain/src/adapters/wikimedia/category-gate.ts
/**
 * Category membership gate — category presence never implies automatic inclusion (BB-045).
 */
import {
  DEFAULT_WIKIMEDIA_CATEGORY_GRAPH,
  listSeedCategoryTitles,
  resolveGraphNodeIdByTitle,
  traverseCategoryGraph,
  type CategoryGraph,
} from './category-graph.js';
import type { WikimediaCategoryGateResult } from './types.js';

export type EvaluateCategoryGateInput = {
  readonly pageCategories: readonly string[];
  readonly graph?: CategoryGraph;
};

/**
 * Fail-closed gate: a page is eligible only when it matches at least one curated seed category.
 * Expand/reference categories may appear in traversal metadata but never grant inclusion alone.
 */
export function evaluateCategoryGate(input: EvaluateCategoryGateInput): WikimediaCategoryGateResult {
  const graph = input.graph ?? DEFAULT_WIKIMEDIA_CATEGORY_GRAPH;
  const seedTitles = new Set(listSeedCategoryTitles(graph).map((title) => title.toLowerCase()));
  const matchedSeedCategories = input.pageCategories.filter((category) =>
    seedTitles.has(category.toLowerCase()),
  );

  const traversedNodeIds = traverseCategoryGraph(graph, input.pageCategories);
  const traversedCategories = traversedNodeIds
    .map((nodeId) => graph.nodes.find((node) => node.id === nodeId)?.title)
    .filter((title): title is string => Boolean(title));

  if (matchedSeedCategories.length === 0) {
    const referenceOnly = input.pageCategories.some((category) => {
      const nodeId = resolveGraphNodeIdByTitle(graph, category);
      if (!nodeId) {
        return false;
      }
      const node = graph.nodes.find((entry) => entry.id === nodeId);
      return node?.role === 'reference' || node?.role === 'expand';
    });

    return {
      passed: false,
      matchedSeedCategories: [],
      traversedCategories,
      reason: referenceOnly
        ? 'Page categories match expand/reference nodes only; seed category required'
        : 'No curated seed category membership',
    };
  }

  return {
    passed: true,
    matchedSeedCategories,
    traversedCategories,
    reason: `Matched seed categories: ${matchedSeedCategories.join(', ')}`,
  };
}

export function assertCategoryGatePassed(result: WikimediaCategoryGateResult): void {
  if (!result.passed) {
    throw new Error(`Wikimedia category gate failed: ${result.reason}`);
  }
}

File: workers/research/src/black_book_research/adapters/wikimedia/category_gate.py
"""Category membership gate — category presence never implies automatic inclusion (BB-045)."""

from __future__ import annotations

from .category_graph import (
    DEFAULT_WIKIMEDIA_CATEGORY_GRAPH,
    CategoryGraph,
    list_seed_category_titles,
    traverse_category_graph,
)
from .types import WikimediaCategoryGateResult


def evaluate_category_gate(
    page_categories: tuple[str, ...],
    *,
    graph: CategoryGraph = DEFAULT_WIKIMEDIA_CATEGORY_GRAPH,
) -> WikimediaCategoryGateResult:
    seed_titles = {title.lower() for title in list_seed_category_titles(graph)}
    matched = tuple(category for category in page_categories if category.lower() in seed_titles)
    traversed_ids = traverse_category_graph(graph, page_categories)
    title_by_id = {node.id: node.title for node in graph.nodes}
    traversed_categories = tuple(title_by_id[node_id] for node_id in traversed_ids if node_id in title_by_id)

    if not matched:
        return WikimediaCategoryGateResult(
            passed=False,
            matched_seed_categories=(),
            traversed_categories=traversed_categories,
            reason="No curated seed category membership",
        )

    return WikimediaCategoryGateResult(
        passed=True,
        matched_seed_categories=matched,
        traversed_categories=traversed_categories,
        reason=f"Matched seed categories: {', '.join(matched)}",
    )


def assert_category_gate_passed(result: WikimediaCategoryGateResult) -> None:
    if not result.passed:
        raise ValueError(f"Wikimedia category gate failed: {result.reason}")
